Sitka-Sound-Science-Center/sitka-landslide
# Integrate real data into detail page

## Overview

Converts the timestamp detail page to use real rainfall observations/forecasts instead of the placeholder data file, and changes the way the data file is generated and loaded.

Most of the data fields needed here were already created for use on the main page, or added in PR #17 in anticipation of needing them here. The exceptions were related to timestamp formatting--this adds a field for the "April 7 · 10AM-1PM" label at the top of the page and one to provide a truncated timestamp for use in the URLs, to avoid cluttering them up with too much noise (i.e. the seconds and timezone offset).

#### Data loading
The more exciting part of this PR relates to how the rainfall data gets loaded and supplied to the page builds.  In PR #17 I moved the script into the app and had it get called in the `getStaticProps` function of the main page.  That worked well for that one page, but it turns out it's not at all easy to share data that's loaded that way on a whole-app level.  The simplest implementation code-wise would have required calling the same function in the `getStaticProps` functions for both pages and also in `getStaticPaths` for the detail page.  And in fact `getStaticProps` gets called once for each item returned by `getStaticPaths`, so that that would have resulted in something like 30 simultaneous calls to the same two endpoints for each build.

So that's not great, and initially I worked around it by leaving the one call in `index.js` but adding a bit there to write the results out to a `rainfall.json` file, which `[timestamp].js` then reads in via import.  That works fine in development, but not for an actual build, because pages are built in parallel, so there's no guarantee (or even really a possibility) that the `rainfall.json` file would be in place by the time the `[timestamp].js` builds try to load it.

So I backtracked on the whole "call it dynamically from the page" thing and instead converted it to a script that just downloads the data and writes it to `rainfall.json`.  Then I added that script to the `build` and `dev` scripts in `package.json`.  So now before the `next` build even starts, the build process gets the data and writes it to disk, so it's there for both pages to load via import.

This also means code changes won't reload the data in development--once you've started the dev server, it'll keep showing the same data until you either stop and restart it or manually run `yarn get-data` to update the file.

Also, I renamed the script file from `rainfall.js` to `getRainfall.js`. Partly to make it more obvious that it's a script, but mainly because it was confusing to have both `rainfall.js` and `rainfall.json`.

#### Other notes

- The rainfall amount is shown in inches, but the historical graph wasn't.  Or rather, the axis label said it was, but the data was still in millimeters.  But I just ran it through a `map()` call right after import to do the conversion.
- There's now nothing using the `index.json` sample file, so I removed it.
- The timestamps in `index.json` ended with the hour.  I made the shortened-for-URL timestamps end with minutes, even though the minutes are always 00.  I didn't actually notice that difference until just now.  It would be easy to change, but I kind of wonder if having the `:00` makes it easier to visually parse the value as a time.
- When I added `EXAGGERATION_FACTOR` in PR #17 to facilitate showing risks besides "Low", I applied it only to the risk calculation.  But that made things look weird, since the rainfall amounts were still way down near the bottom of the graph.  So I moved it so that it applies to rainfall amounts, which then affects the risk calculation, so everything's consistent.

Resolves #14 

### Demo

![image](https://user-images.githubusercontent.com/6598836/162255242-9cbc8458-ba16-4273-972e-dbd640f416b7.png)

![image](https://user-images.githubusercontent.com/6598836/162255268-72ffc958-a568-461d-9201-aa7758370b1c.png)

## Testing Instructions

- Set `EXAGGERATION_FACTOR` in `frontend/data/getRainfall.js` to something highish.
- Run `./scripts/server`
- Go to [the main view](http://localhost:3008/), click into the accordion, and open some of the detail pages.
- Also click the header to get to http://localhost:3008/detail/current
- Everything should look sensible and correct.
- In the `frontend/` directory, run `yarn build` then `yarn start`.  That will build the static site then start a "production mode" server.
- Go to http://localhost:3000/ and click around. It should look just the same as what you saw from the dev server.


## Checklist
